### PR TITLE
HDDS-8580. Reduce memory usage in ContainerKeyMapperTask#reprocess

### PR DESCRIPTION
**Pull request by ivandika3**
## What changes were proposed in this pull request?

While setting up for OM performance test during Recon OM full snapshot, I removed the Recon DB directory before restarting Recon to trigger full snapshot (essentially bootstrapping a new Recon).

However, it is found after OM DB is successfully downloaded, during the reprocess of ContainerKeyMapperTask, the Recon heap usage increased significantly for large keys table (our cluster has around 350 million keys).

It is found that the issue was caused due to in-memory maps that store all the OM keys during the reprocess. This is a regression introduced in [HDDS-6783](https://issues.apache.org/jira/browse/HDDS-6783). In essence, the patch is to revert the implementation of [HDDS-6783](https://issues.apache.org/jira/browse/HDDS-6783) ONLY for ContainerKeyMapperTask#reprocess.

- ContainerKeyMapperTask#process should not increase the heap memory significantly since the number of delta updates are already limited by the Recon configurations
- [HDDS-6783](https://issues.apache.org/jira/browse/HDDS-6783) aims for atomicity during the Recon OM task updates. However since ContainerKeyMapperTask#reprocess truncate all the Recon Container DB before it starts and rebuilt the Recon Container DB, I think it's acceptable.

After the patch is applied, the Recon heap size stays stable during the full snapshot.

Any suggestion for better approach is welcomed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8580

## How was this patch tested?

Manual test. 

Attached is Recon heap memory before and after the patch.

image

### CODE REVIEW

**Comment by ivandika3** [comment]
@dombizita @smengcl Could you help take a look?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 1]
```diff
@@ -383,4 +369,55 @@ private void handlePutOMKeyEvent(String key, OmKeyInfo omKeyInfo,
...
+            containerKeyPrefix) == 0) {
+          // Save on writes. No need to save same container-key prefix
+          // mapping again.
+          reconContainerMetadataManager.storeContainerKeyMapping(
+              containerKeyPrefix, 1);
+
+          // check if container already exists and
+          // increment the count of containers if it does not exist
+          if (!reconContainerMetadataManager.doesContainerExists(containerId)) {
+            containerCountToIncrement++;
```
Comment (ashishkumar50) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:401]: I think in earlier case containerKeyMap keep appending and increasing memory during reprocess for each key.
After this change we are querying from rocksdb always, I think better we can avoid that.
May be we can keep map in this function locally to keep containerId so that we can avoid querying rocksdb again for same container. There won't be many containers per key and hence memory will not increase much.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 2]
```diff
@@ -383,4 +369,55 @@ private void handlePutOMKeyEvent(String key, OmKeyInfo omKeyInfo,
...
+          }
+
+          // update the count of keys for the given containerID
+          long keyCount = reconContainerMetadataManager
+              .getKeyCountForContainer(containerId);
```
Comment (ashishkumar50) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:406]: getKeyCountForContainer also we can avoid querying from rocksdb always, Initialize keyCount as 0 and if container doesn't exist we can avoid calling this.

**Review by ashishkumar50 — COMMENTED** [review]
Hi @ivandika3, Thanks for working on this. Please find my comments inline.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 1 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:401]: Thank you for the review. Since `containerKeyMap` is the main culprit, maybe we can reuse the `containerKeyCountMap` introduced in HDDS-6783, accumulate it during the whole reprocess and persist it at the end as a batch by reusing the `writeToTheDB` function (`containerKeyMap` argument can be left as an empty map).

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 2 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:406]: Following the previous conversation, we can follow the implementation in `handlePutOMKeyEvent` function using the `containerKeyCountMap` instead, as follows:

```
// update the count of keys for the given containerID
long keyCount;
if (containerKeyCountMap.containsKey(containerId)) {
  keyCount = containerKeyCountMap.get(containerId);
} else {
  keyCount = reconContainerMetadataManager
      .getKeyCountForContainer(containerId);
}

// increment the count and update containerKeyCount.
// keyCount will be 0 if containerID is not found. So, there is no
// need to initialize keyCount for the first time.
containerKeyCountMap.put(containerId, ++keyCount);
```

If the container has been encountered before, we can use the look up from `containerKeyCountMap`.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 1 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:401]: Using `containerKeyCountMap`:
```
if (!containerKeyCountMap.containsKey(containerId)) {
  containerCountToIncrement++;
}
```

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 2 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: @ashishkumar50 I have addressed the suggestion, PTAL, thank you.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 1 above)
Comment (ashishkumar50) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:439]: Yes makes sense.

Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:439]: I have tested the new patch, there is no regression and there was significant improvement `ContainerKeyMapperTask#reprocess` processing time.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 3]
```diff
@@ -383,4 +398,60 @@ private void handlePutOMKeyEvent(String key, OmKeyInfo omKeyInfo,
...
+      for (OmKeyLocationInfo omKeyLocationInfo : omKeyLocationInfoGroup
+          .getLocationList()) {
+        long containerId = omKeyLocationInfo.getContainerID();
+        ContainerKeyPrefix containerKeyPrefix = new ContainerKeyPrefix(
+            containerId, key, keyVersion);
+        if (reconContainerMetadataManager.getCountForContainerKeyPrefix(
+            containerKeyPrefix) == 0) {
+          // Save on writes. No need to save same container-key prefix
+          // mapping again.
+          reconContainerMetadataManager.storeContainerKeyMapping(
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:429]: Thanks for the patch @ivandika3 , though I have one suggestion. After this change, we are storing containerKeyMapping info with every key which is hitting DB everytime, so if we want to improve memory footprint due to containerKeyMap, we can use a threshold and till that threshold number of keys, retain in memory, lets say 1000 keys, and then use batchStore function for containerKeyMap instead of removing it completely. We have done like this in other aync tasks like NSSummary, You can refer that. This way we'll avoid 1 billion DB calls with 1 billion keys.

Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: @devmadhuu Thank you for the review. I have added configuration and support for incremental batching for containerKeyMapping. Please take another look.

Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: @devmadhuu The threshold might be too low, could you help suggest a good default? Or maybe Ishould follow the NSSummary configuration (i.e. 150,000)?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 4]
```diff
@@ -383,4 +445,63 @@ private void handlePutOMKeyEvent(String key, OmKeyInfo omKeyInfo,
...
+    for (OmKeyLocationInfoGroup omKeyLocationInfoGroup : omKeyInfo
+        .getKeyLocationVersions()) {
+      long keyVersion = omKeyLocationInfoGroup.getVersion();
+      for (OmKeyLocationInfo omKeyLocationInfo : omKeyLocationInfoGroup
+          .getLocationList()) {
+        long containerId = omKeyLocationInfo.getContainerID();
+        ContainerKeyPrefix containerKeyPrefix = new ContainerKeyPrefix(
+            containerId, key, keyVersion);
+        if (reconContainerMetadataManager.getCountForContainerKeyPrefix(
+            containerKeyPrefix) == 0) {
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:477]: we can avoid the put operation by adding back the condition "`!containerKeyMap.containsKey(containerKeyPrefix)`"

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 5]
```diff
@@ -120,14 +145,42 @@ public Pair<String, Boolean> reprocess(OMMetadataManager omMetadataManager) {
...
       LOG.error("Unable to write Container Key Prefix data in Recon DB.", e);
       return new ImmutablePair<>(getTaskName(), false);
     }
     return new ImmutablePair<>(getTaskName(), true);
   }
 
+  private boolean flushAndCommitContainerKeyMapToDB(
+      Map<ContainerKeyPrefix, Integer> containerKeyMap) {
+    try {
+      writeToTheDB(containerKeyMap, Collections.emptyMap(),
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:164]: I think, we should flush the "`containerKeyCountMap`" also and optimize more memory footprint if we are introducing this flush mechanism, so you can refactor the code a bit and pass "`containerKeyCountMap`" instead of sending empty map in 2nd argument.

**Review by devmadhuu — COMMENTED** [review]
Added few comments, pls check

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 4 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: Thank you. Addressed.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 5 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: Thank you for the suggestion. I have included `containerKeyCountMap`  into the flush, and removing the last `writeToDB` in the 'reprocess' function. I have also moved the flush function to an outer scope to prevent inaccurate container key info when there are multiple bucket layouts.

Comment (devmadhuu) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: Thanks @ivandika3 for fixing comments. Patch LGTM. +1

**Comment by ivandika3** [comment]
CI failure does not seem to be related.

**Comment by ivandika3** [comment]
Hi @dombizita could you help review this?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 6]
```diff
@@ -383,4 +443,64 @@ private void handlePutOMKeyEvent(String key, OmKeyInfo omKeyInfo,
...
+   * @param key key String
+   * @param omKeyInfo omKeyInfo value
+   * @param containerKeyMap we keep the added containerKeys in this map
+   *                        to allow incremental batching to containerKeyTable
+   * @param containerKeyCountMap we keep the containerKey counts in this map 
+   *                             to allow batching to containerKeyCountTable
+   *                             after reprocessing is done
+   * @throws IOException if unable to write to recon DB.
+   */
+  private void handleKeyReprocess(String key,
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:459]: I just want to make sure I understand this method correctly: the only difference between `handlePutOMKeyEvent` and this is that we don't ask from the `reconContainerMetadataManager` if the container exists, we rather ask if the `containerKeyCountMap` contains the container (as a key) and if it does not have it we just easily set the key count to zero (instead of asking it from the `reconContainerMetadataManager`), right?

**Review by dombizita — COMMENTED** [review]
thanks for working on this @ivandika3, I only had one question about a method (to make sure that I understand it correctly) and one about a comment change, beside this it looks good to me!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 6 above)
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:459]: Thank you for the review.
> if it does not have it we just easily set the key count to zero

If the container does not exist during the reprocess (when `handleKeyReprocess` is called), it won't be set in the `containerKeyCountMap` at all and therefore won't be persisted to the Recon container DB. I think the comment "keyCount will be 0 if containerID is not found" in `handleKeyReprocess` is a bit misleading since it was duplicated from `handlePutOMKeyEvent` where `ReconContainerMetadataManager#getKeyCountForContainer` will return 0 if the container is not there.  I will remove it to clear things up.

> we don't ask from the reconContainerMetadataManager if the container exists, we rather ask if the containerKeyCountMap

This is correct.

To summarize, the major differences between `handlePutOMKeyEvent` and `handleKeyReprocess` are:

1. It doesn't need to keep track `deletedContainerKeyList` since reprocess will only insert (PUT) container info
2. All the `(containerId -> key count)` related operation are done in-memory (no need to access `containerKeyCountTable`). It is only flushed after the whole reprocess is done.

Hope this clarifies.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 7]
```diff
@@ -76,13 +87,12 @@ public ContainerKeyMapperTask(ReconContainerMetadataManager
   @Override
   public Pair<String, Boolean> reprocess(OMMetadataManager omMetadataManager) {
     long omKeyCount = 0;
-    // Maps the (container, key) -> count
+
+    // In-memory maps for fast look up and batch write
+    // key -> containerId
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: Thank you for the review. Yes, you are correct. Thanks for the catch. Addressed.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java [hunk 8]
```diff
@@ -145,8 +196,17 @@ public Pair<String, Boolean> process(OMUpdateEventBatch events) {
     Iterator<OMDBUpdateEvent> eventIterator = events.getIterator();
     int eventCount = 0;
     final Collection<String> taskTables = getTaskTables();
+
+    // In-memory maps for fast look up and batch write
+    // (HDDS-8580) containerKeyMap map is allowed to be used
+    // in "process" without batching since the maximum number of keys
+    // is bounded by delta limit configurations
+
+    // key -> containerId
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: Addressed. Thank you.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 6 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:459]: thanks for the detailed explanation @ivandika3, it makes sense!

**Review by dombizita — APPROVED** [review]

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 3 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: @ivandika3 @devmadhuu what was the conclusion here?

**Comment by ferhui** [comment]
@ashishkumar50 do you have other comments? Plan to merge this PR. Thanks

**Comment by dombizita** [comment]
> @ashishkumar50 do you have other comments? Plan to merge this PR. Thanks

I'm waiting to get a reply on [this](https://github.com/apache/ozone/pull/4696#discussion_r1192175162) comment.

cc @devmadhuu @ivandika3

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java (hunk 3 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java]: thanks @ivandika3 for the testing and updating the patch!

**Review by ashishkumar50 — COMMENTED** [review]
@ivandika3, Thanks for updating the patch, LGTM +1.

**Comment by dombizita** [comment]
I just wanted to merge this PR but there is a merge conflict, can you resolve it @ivandika3?

**Comment by ivandika3** [comment]
Hi @dombizita, I have resolved the conflict. Thank you.

**Comment by dombizita** [comment]
thanks for the patch @ivandika3! thanks for the review @devmadhuu @ashishkumar50!

**Comment by whbing** [comment]
mvn build failed in https://github.com/apache/ozone/actions/runs/5162988533, also failed in master branch now.
If it's related to this PR? Thanks !

```
Error:  /home/runner/work/ozone/ozone/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperTask.java:[472,49] 
org.apache.hadoop.ozone.recon.api.types.ContainerKeyPrefix is abstract; cannot be instantiated
```

**Comment by ivandika3** [comment]
Hi @whbing, there is an incompatibility due to refactoring done on HDDS-8733. I have raised https://github.com/apache/ozone/pull/4826 to include the the fix for the incompatibility. @dombizita @szetszwo Could you help to resolve the conflict? Sorry for the inconvenience.

**Comment by dombizita** [comment]
thanks for finding this @whbing, #4826 is merged. thanks for the quick fix @ivandika3!
